README.md
# file-domain-server

![version](https://img.shields.io/github/v/tag/johanbook/file-domain-server)
![docker pulls](https://img.shields.io/docker/pulls/johanbook/file-domain-server)
![vulnerabilities](https://img.shields.io/snyk/vulnerabilities/github/johanbook/file-domain-server)

**file-domain-server** is a simplistic Nodejs HTTP file server that serves
different folders depending on the HTTP `Host` header. It is mainly intended to
run on an internal network with its own DNS server.

The project aims to be minimal production dependency.

## Get started

The server can be run either via a Nodejs runtime or through Docker.

### Folder structure

Create an asset folder, named `build` by default. In this folder create one
folder for each domain that should be served. These folders should contain the
files served on the respective domains. For example:

```sh
build
build/localhost
build/localhost/favicon.ico
build/localhost/index.html
build/mybusiness.com
build/mybusiness.com/favicon.ico
build/mybusiness.com/index.html
build/mybusiness.com/script.js
build/mybusiness.com/styles.css
build/mycat.com
build/mycat.com/favicon.ico
build/mycat.com/index.html
build/mycat.com/script.js
```

### Using Nodejs

Install and run the server using

```sh
npm ci --production
ROOT_FILE_PATH=/path/to/folder node src
```

### Using Docker

The server can be run through Docker by mounting in the served directory as a
volume. To the serve the `build` folder in the current directory, run

```sh
docker run \
	--detach \
	--publish 8080:8080 \
	--restart unless-stopped \
	--volume ${PWD}/build/:/app/build/ \
	johanbook/file-domain-server:latest
```

The Docker images are available for multiple architectures, including armv7.

## Configuration

The following can be configured as environment variables:

- **ACCESS_LOGS** if access logs are showed (default `false`).
- **DEFAULT_INDEX_FILE** name of file served on paths ending with a trailing
  slash (default `index.html`).
- **FALLBACK_FILE** name of file served as a fallback if requested file is not
  found. Set to `index.html` to enable client-side routing or to e.g.
  `error.html` to display an error page (note that response still returns a 200
  status code). It is undefined by default and no fallback file will be served.
- **LOG_LEVEL** used log level. Available values are `debug`, `info`, `warning`
  and `error` (default `info`).
- **PORT** the port the server should listen on (default 8080).
- **ROOT_FILE_PATH** root directory of served file tree (default `build`).

## Documentation

For documentation on technologies, security aspect and similar, see the
[contributions file](./CONTRIBUTING.md).

## Contributing

For contributions and development procedures, see the
[contributions file](./CONTRIBUTING.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/johanbook/file-domain-server/compare/v1.0.0...v1.0.1) (2022-10-18)

## 1.0.0 (2022-10-18)


src/index.js
const fs = require("node:fs");
const http = require("node:http");
const path = require("node:path");

const mime = require("mime");

const logger = require("./logger");

const DEFAULT_INDEX_FILE = process.env.DEFAULT_INDEX_FILE || "index.html";
const FALLBACK_FILE = process.env.FALLBACK_FILE || null;
const PORT = process.env.PORT || 8080;
const ROOT_FILE_PATH = process.env.ROOT_FILE_PATH || "build";

function getFile(filePath) {
  try {
    const file = fs.readFileSync(filePath);
    return file;
  } catch {
    logger.error(`Could not find file "${filePath}"`);
    return;
  }
}

function getFileMime(filePath) {
  return mime.getType(filePath);
}

function getDomainFromRequest(req) {
  const headers = req.headers;
  const host = headers.host;
  const domain = host.split(":")[0];
  return domain;
}

function getFilePathFromRequest(req) {
  return req.url.split("?")[0];
}

function createFilePath(domain, filePath) {
  if (filePath === "/") {
    filePath = DEFAULT_INDEX_FILE;
  }
  return path.join(ROOT_FILE_PATH, domain, filePath);
}

function handleRequest(req, res) {
  const domain = getDomainFromRequest(req);
  const filePath = getFilePathFromRequest(req);
  logger.debug(`Requesting ${domain}${filePath}`);

  const actualFilePath = createFilePath(domain, filePath);
  let file = getFile(actualFilePath);

  if (!file && FALLBACK_FILE) {
    const fallbackFilePath = createFilePath(domain, FALLBACK_FILE);
    logger.debug(`Serving fallback file: ${fallbackFilePath}`);
    file = getFile(fallbackFilePath);
  }

  if (!file) {
    res.writeHead(404);
    res.end();
    logger.accessLog(req, res);
    return;
  }

  const mimeType = getFileMime(actualFilePath);

  res.setHeader("Content-Type", mimeType);
  res.writeHead(200);
  res.end(file);
  logger.accessLog(req, res);
}

const server = http.createServer(handleRequest);
server.listen(PORT);

logger.info(`Listening on port ${PORT}`);


src/logger.js
const ACCESS_LOGS = process.env.ACCESS_LOGS || false;
const LOG_LEVEL = process.env.LOG_LEVEL || "info";

const LOG_LEVELS = {
  DEBUG: 1,
  INFO: 2,
  WARNING: 3,
  ERROR: 4,
};

const LEVEL = LOG_LEVELS[LOG_LEVEL.toUpperCase()];

exports.debug = function (message) {
  if (LEVEL <= LOG_LEVELS.DEBUG) {
    console.debug(`DEBUG: ${message}`);
  }
};

exports.info = function (message) {
  if (LEVEL <= LOG_LEVELS.INFO) {
    console.info(`INFO: ${message}`);
  }
};

exports.warn = function (message) {
  if (LEVEL <= LOG_LEVELS.WARNING) {
    console.warn(`WARNING: ${message}`);
  }
};

exports.error = function (message) {
  if (LEVEL <= LOG_LEVELS.ERROR) {
    console.error(`ERROR: ${message}`);
  }
};

exports.accessLog = function (req, res) {
  if (ACCESS_LOGS) {
    console.info(`${req.headers.host}${req.url} ${res.statusCode}`);
  }
};


src/server.js
const fs = require("node:fs");
const http = require("node:http");
const path = require("node:path");

const mime = require("mime");

const logger = require("./logger");

const DEFAULT_INDEX_FILE = process.env.DEFAULT_INDEX_FILE || "index.html";
const FALLBACK_FILE = process.env.FALLBACK_FILE || null;
const PORT = process.env.PORT || 8080;
const ROOT_FILE_PATH = process.env.ROOT_FILE_PATH || "build";

function getFile(filePath) {
  try {
    const file = fs.readFileSync(filePath);
    return file;
  } catch {
    logger.error(`Could not find file "${filePath}"`);
    return;
  }
}

function getFileMime(filePath) {
  return mime.getType(filePath);
}

function getDomainFromRequest(req) {
  const headers = req.headers;
  const host = headers.host;
  const domain = host.split(":")[0];
  return domain;
}

function getFilePathFromRequest(req) {
  return req.url.split("?")[0];
}

function createFilePath(domain, filePath) {
  if (filePath === "/") {
    filePath = DEFAULT_INDEX_FILE;
  }
  return path.join(ROOT_FILE_PATH, domain, filePath);
}

function handleRequest(req, res) {
  const domain = getDomainFromRequest(req);
  const filePath = getFilePathFromRequest(req);
  logger.debug(`Requesting ${domain}${filePath}`);

  const actualFilePath = createFilePath(domain, filePath);
  let file = getFile(actualFilePath);

  if (!file && FALLBACK_FILE) {
    const fallbackFilePath = createFilePath(domain, FALLBACK_FILE);
    logger.debug(`Serving fallback file: ${fallbackFilePath}`);
    file = getFile(fallbackFilePath);
  }

  if (!file) {
    res.writeHead(404);
    res.end();
    logger.accessLog(req, res);
    return;
  }

  const mimeType = getFileMime(actualFilePath);

  res.setHeader("Content-Type", mimeType);
  res.writeHead(200);
  res.end(file);
  logger.accessLog(req, res);
}

exports.createServer = function (port = PORT) {
  const server = http.createServer(handleRequest);
  server.listen(port);

  logger.info(`Listening on port ${port}`);
  return server;
};
